File: app/src/main/java/com/falcon/restaurants/screens/restaurant/RestaurantsListAdapter.java
package com.falcon.restaurants.screens.restaurant;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import androidx.recyclerview.widget.RecyclerView;
import com.falcon.restaurants.R;
import com.falcon.restaurants.room.restaurant.Restaurant;
import com.falcon.restaurants.screens.common.ImageLoader;
import com.falcon.restaurants.utils.Constants;
import java.util.List;

public class RestaurantsListAdapter extends RecyclerView.Adapter<RestaurantsListAdapter.ModelViewHolder> {

    String TAG = "RestaurantsListAdapter" ;
    private List<Restaurant> main_arrlst;

    Context context ;
    private final LayoutInflater layoutInflater;
    private FilterRestaurantsUseCase filterRestaurantsUseCase;
    private ImageLoader imageLoader;

    private OnRestaurantClickListener onRestaurantClickListener;

    public interface OnRestaurantClickListener {
        void onRestaurantClicked(String typeId);
    }

    public RestaurantsListAdapter(Context context,
                                  LayoutInflater layoutInflater,
                                  FilterRestaurantsUseCase filterRestaurantsUseCase,
                                  ImageLoader imageLoader) {
        this.context = context;
        this.layoutInflater = layoutInflater ;
        this.filterRestaurantsUseCase = filterRestaurantsUseCase;
        this.imageLoader = imageLoader;
    }

    class ModelViewHolder extends RecyclerView.ViewHolder {
        private final TextView name_txt;
        private final ImageView restaurant_img;

        private ModelViewHolder(View itemView) {
            super(itemView);
            name_txt = itemView.findViewById(R.id.name_txt);
            restaurant_img = itemView.findViewById(R.id.restaurant_img);

            itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    String typeId = (String) name_txt.getTag();
                    onRestaurantClickListener.onRestaurantClicked(typeId);
                }
            });
        }
    }

    @Override
    public ModelViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View itemView = layoutInflater.inflate(R.layout.recyclerview_restaurant_item, parent, false);
        return new ModelViewHolder(itemView);
    }

    @Override
    public void onBindViewHolder(ModelViewHolder holder, int position) {
        if (main_arrlst != null) {
            Restaurant current = main_arrlst.get(position);
            holder.name_txt.setText(current.getName());
            holder.name_txt.setTag(current.getId());

            imageLoader.loadImage(holder.restaurant_img,
                    Constants.imagesCategoriesUrl + current.getImageUrl());
        }
    }

    public void setMainList(List<Restaurant> main_arrlst){
        if(main_arrlst != null) {
            this.main_arrlst = main_arrlst;
            filterRestaurantsUseCase.setList(main_arrlst);
            notifyDataSetChanged();
        }
    }

    @Override
    public int getItemCount() {
        if (main_arrlst != null)
            return main_arrlst.size();
        else return 0;
    }

    public void filter(String queryText){
        main_arrlst = filterRestaurantsUseCase.filter(queryText);
        notifyDataSetChanged();
    }

    public void setOnRestaurantClickListener(OnRestaurantClickListener onRestaurantClickListener) {
        this.onRestaurantClickListener = onRestaurantClickListener;
    }
}

File: app/src/main/java/com/falcon/restaurants/screens/restaurant/RestaurantsListAdapter.kt
package com.falcon.restaurants.screens.restaurant

import android.content.Context
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.falcon.restaurants.databinding.RecyclerviewRestaurantItemBinding
import com.falcon.restaurants.room.restaurant.Restaurant
import com.falcon.restaurants.screens.common.ImageLoader
import com.falcon.restaurants.utils.Constants

class RestaurantsListAdapter (
    val context: Context,
    val layoutInflater: LayoutInflater,
    val filterRestaurantsUseCase: FilterRestaurantsUseCase,
    val imageLoader: ImageLoader
    ): RecyclerView.Adapter<RestaurantsListAdapter.ModelViewHolder>() {

    val TAG: String = "RestaurantsListAdapter" 

    var main_arrlst = mutableListOf<Restaurant>()
        set(value){
            field = value
            notifyDataSetChanged()
        }

    lateinit var onRestaurantItemClicked: (restaurantId: String) -> Unit

    inner class ModelViewHolder(private val binding: RecyclerviewRestaurantItemBinding) : RecyclerView.ViewHolder(binding.root) {

        lateinit var restaurantId: String

        fun bind(currentRestaurant: Restaurant){
            binding.nameTxt.text = currentRestaurant.name
            restaurantId = currentRestaurant.id
            imageLoader.loadImage(binding.restaurantImg, Constants.imagesCategoriesUrl + currentRestaurant.imageUrl)
        }

        init {
            itemView.setOnClickListener { view -> onRestaurantItemClicked(restaurantId) }
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ModelViewHolder {
        val binding = RecyclerviewRestaurantItemBinding.inflate(layoutInflater, parent, false)
        return ModelViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ModelViewHolder, position: Int) {
        if (main_arrlst != null) {
            holder.bind(main_arrlst.get(position))
        }
    }

    override fun getItemCount(): Int {
        if (main_arrlst != null)
            return main_arrlst.size
        else return 0
    }

    fun filter(queryText: String){
        main_arrlst = filterRestaurantsUseCase.filter(queryText)
        notifyDataSetChanged()
    }

    fun setList(restaurants: List<Restaurant>?) {
        main_arrlst = restaurants as MutableList<Restaurant>
        filterRestaurantsUseCase.setList(restaurants)
    }

}